File: c++/Solution297.cpp
//DFS serialize(), deserialize(): O(V) O(V)
class Codec {//Solution297
public:
    string serialize(TreeNode* root) {
        ostringstream oss;
        serialize(root, oss);
        return oss.str();
    }

    void serialize(TreeNode* root, ostringstream& oss) {
        if (!root) {
            oss << "N ";
        } else {
            oss << root->val << " ";
            serialize(root->left, oss);
            serialize(root->right, oss);
        }
    }

    TreeNode* deserialize(string data) {
        istringstream iss(data);
        return deserialize(iss);
    }

    TreeNode* deserialize(istringstream& iss) {
        string val;
        iss >> val;
        if (val == "N") {
            return nullptr;
        } else {
            TreeNode* node = new TreeNode(stoi(val));
            node->left = deserialize(iss);
            node->right = deserialize(iss);
            return node;
        }
    }
};


//DFS serialize(), deserialize(): O(V) O(V)
class Codec {//Solution297_2
public:
    string serialize(TreeNode* root) {
        if (!root) return "N";

        return format("{},{},{}", root->val, serialize(root->left), serialize(root->right));
    }

    TreeNode* deserialize(string data) {
        queue<string> q; // 有可能是兩位數以上, 用string
        string tmp = "";
        for (char ch : data) {
            if (ch == ',') {
                q.push(tmp);
                tmp = "";
            } else {
                tmp.push_back(ch);
            }
        }
        q.push(tmp); // 放入最後
        return buildTree(q);
    }

    TreeNode* buildTree(queue<string>& q) { // 依照preorder解析
        string val = q.front(); q.pop();
        if (val == "N") return nullptr;

        TreeNode* node = new TreeNode(stoi(val));
        node->left = buildTree(q);
        node->right = buildTree(q);
        return node;
    }
};

File: java/Solution297.java
package leetCode.java;

import java.util.*;

//DFS serialize(), deserialize(): O(V) O(V)
class Codec {//Solution297

    // Encodes a tree to a single string.
    public String serialize(TreeNode root) {
        //Pre-order前序遍歷, root, left, right
        if (root == null) {
            return "N";
        }
        return root.val + "," + serialize(root.left) + "," + serialize(root.right);
    }

    // Decodes your encoded data to tree.
    public TreeNode deserialize(String data) {
        String[] vals = data.split(",");
        Deque<String> q = new ArrayDeque<>(Arrays.asList(vals));
        return dfs(q); 
    }

    public TreeNode dfs(Deque<String> q) {
        String val = q.poll();
        if ("N".equals(val)) {
            return null;
        }

        TreeNode node = new TreeNode(Integer.parseInt(val));
        node.left = dfs(q);
        node.right = dfs(q);
        return node;
    }
}
/*
        1
    2       3
  4   5   N   6
 N N N N     N N

Pre-order:
1,2,4,N,N,5,N,N,3,N,6,N,N
*/